oio/xcute/orchestrator.py

```diff
@@ -375,77 +375,144 @@ def process_reply(self, beanstalkd_job_id, encoded_reply):
 
         yield None
 
-    def refresh_all_beanstalkd(self):
+    def refresh_beanstalkd_workers_forever(self):
         """
-            Get all the beanstalkd and their tubes
+        Refresh beanstalkd workers by looking at the score,
+        existing tubes and tube statistics.
         """
-
         while self.running:
-            all_beanstalkd = self.conscience_client.all_services('beanstalkd')
-
-            all_beanstalkd_with_tubes = {}
-            for beanstalkd in all_beanstalkd:
-                beanstalkd_addr = beanstalkd['addr']
-
-                try:
-                    beanstalkd_tubes = self.get_beanstalkd_tubes(
-                        beanstalkd_addr)
-                except ConnectionError:
-                    continue
-
-                all_beanstalkd_with_tubes[beanstalkd_addr] = (
-                    beanstalkd, beanstalkd_tubes)
+            try:
+                beanstalkd_workers = self._find_beanstalkd_workers()
+            except Exception as exc:
+                self.logger.error(
+                    'Fail to find beanstalkd workers: %s', exc)
+                beanstalkd_workers = dict()
```

Comment:
Isn't it a bit too much? We could keep trying to send jobs to the beanstalkd we already found. If there's really a problem, they will be marked as broken anyway.

Reply:
The only errors that can occur here are conscience errors. If there is a problem with the conscience, it is better to wait until the problem is resolved before for the good progress of the job.

Reply:
Why? We don't need the conscience to complete the job, do we?

Reply:
Most jobs indirectly use the conscience (to know the address of service, to distribute the data on rawx, ...). But actually, we can consider these errors will be listed in the job information. In addition, there may be this problem of communication with conscience only for the orchestrator.

Reply:
However, to continue sending tasks to a beanstalkd, we need to be sure that it still exists and has a positive score. If we don't do this verification, we could send tasks to a beanstalkd which should no longer receive them.

Reply:
I suggest you add a comment to improve this point later.